controller/src/main/java/org/jboss/as/controller/security/CredentialReference.java

```diff
@@ -169,12 +168,21 @@ private static AttributeMarshaller credentialReferenceAttributeMarshaller() {
             @Override
             public void marshallAsElement(AttributeDefinition attribute, ModelNode credentialReferenceModelNode, boolean marshallDefault, XMLStreamWriter writer) throws XMLStreamException {
                 writer.writeStartElement(attribute.getXmlName());
-                if (credentialReferenceModelNode.hasDefined(clearTextAttribute.getName())) {
-                    clearTextAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
+                ModelNode crNode = null;
+                if (credentialReferenceModelNode.hasDefined(attribute.getName())) {
+                    crNode = credentialReferenceModelNode.get(attribute.getName());
                 } else {
-                    credentialStoreAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
-                    credentialAliasAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
-                    credentialTypeAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
+                    crNode = credentialReferenceModelNode;
```

Comment:
Note I have assigned a JIRA to @ctomc to polish up the javadoc on AttributeMarshaller. The marshallAsElement javadoc is missing, although marshalAsAttribute gives a good indication of the expected behavior.